dbt/adapters/bigquery/connections.py

```diff
@@ -452,6 +481,18 @@ def get_labels_from_query_comment(cls):
 
         return {}
 
+    def generate_job_id(self) -> str:
+        # Generating a fresh job_id for every _query_and_results call to avoid job_id reuse.
+        # Generating a job id instead of persisting a BigQuery-generated one after client.query is called.
+        # Using BigQuery's job_id can lead to a race condition if a job has been started and a termination
+        # is sent before the job_id was stored, leading to a failure to cancel the job.
+        # By predetermining job_ids (uuid4), we can persist the job_id before the job has been kicked off.
+        # Doing this, the race condition only leads to attempting to cancel a job that doesn't exist.
+        job_id = str(uuid.uuid4())
+        thread_id = self.get_thread_identifier()
+        self.jobs_by_thread[thread_id] = self.jobs_by_thread.get(thread_id, []) + [job_id]
```

Comment:
This would be cleaner with `defaultdict`:
```python
def __init__(...):
    super().__init__(...)
    self.jobs_by_thread = collections.defaultdict(list)

def generate_job_id(self):
    job_id = str(uuid.uuid4())
    thread_id = self.get_thread_identifer()
    self.jobs_by_thread[thread_id].append(job_id)
    return job_id
```

Reply:
good idea, thanks!